### PR TITLE
Roothash processing fees

### PR DESCRIPTION
**Pull request by kostko**
Part of #2406.

TODO
* [x] Charge per-epoch maintenance fee.
* [x] Suspend runtimes which did not pay the maintenance fee.
* [x] Restore suspended runtimes from genesis dump.
* [x] Resume runtimes which paid the maintenance fee.
* [x] Emit runtime registered event when the runtime is resumed.
* [x] Add missing existence checks to `registerRuntime`.
* [x] E2E tests.

### CODE REVIEW

File: go/consensus/tendermint/apps/registry/genesis.go [hunk 1]
```diff
@@ -75,6 +75,25 @@ func (app *registryApplication) InitChain(ctx *abci.Context, request types.Reque
...
+			"runtime_owner", v.Signature.PublicKey,
+		)
+		if err := app.registerRuntime(ctx, state, v); err != nil {
+			app.logger.Error("InitChain: failed to register runtime",
+				"err", err,
+				"runtime", v,
+			)
+			return errors.Wrap(err, "registry: genesis suspended runtime registration failure")
+		}
+		var rt registry.Runtime
```
Comment (Yawning) [inline go/consensus/tendermint/apps/registry/genesis.go]: Instead of doing this, just make `app.registerRuntime` return `(*registry.Runtime, error)`?

Otherwise, the `cbor.Unmarshal` call failing is a sign that something when absolutely horrifically wrong since the runtime registration does an Unmarshal.

File: go/consensus/tendermint/apps/registry/state/state.go [hunk 2]
```diff
@@ -238,82 +242,157 @@ func (s *ImmutableState) SignedNodes() ([]*node.SignedNode, error) {
...
+}
+
+func (s *ImmutableState) getRuntime(keyFmt *keyformat.KeyFormat, id common.Namespace) (*registry.Runtime, error) {
+	signedRuntime, err := s.getSignedRuntime(keyFmt, id)
+	if err != nil {
 		return nil, err
 	}
 	var runtime registry.Runtime
-	if err = cbor.Unmarshal(signedRuntime.Blob, &runtime); err != nil {
+	if err := cbor.Unmarshal(signedRuntime.Blob, &runtime); err != nil {
```
Comment (Yawning) [inline go/consensus/tendermint/apps/registry/state/state.go]: Why the change here? `err` is already declared and the correct type...

File: go/consensus/tendermint/apps/registry/state/state.go [hunk 3]
```diff
@@ -238,82 +242,157 @@ func (s *ImmutableState) SignedNodes() ([]*node.SignedNode, error) {
...
+		}
+		runtimes = append(runtimes, &rt)
+	})
+
+	return runtimes, nil
+}
+
+// AllRuntimes returns a list of all registered runtimes.
+//
+// This excludes any suspended runtimes.
```
Comment (Yawning) [inline go/consensus/tendermint/apps/registry/state/state.go]: The comment does not match what the function does.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 4]
```diff
@@ -225,6 +228,56 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
...
+			"new_node", newNode,
+			"existing_node", existingNode,
+			"entity", newNode.EntityID,
+		)
+		return registry.ErrInvalidArgument
+	}
+
+	// For each runtime the node registers for, require it to pay a maintenance fee for
+	// each epoch the node is registered in.
+	additionalEpochs := newNode.Expiration - uint64(epoch)
```
Comment (Yawning) [inline go/consensus/tendermint/apps/registry/transactions.go]: It is possible to create a new node, that expires in the current epoch (additionalEpochs = 0), without paying gas.  I don't think the node will actually do anything, but that doesn't seem great.

File: go/oasis-test-runner/scenario/e2e/identity_cli.go [hunk 5]
```diff
@@ -54,7 +55,7 @@ func (i *identityCLIImpl) Run(childEnv *env.Env) error {
 		"identity", "init",
 		"--" + common.CfgDataDir, i.dataDir,
 	}
-	if err := runSubCommand(childEnv, "identity-init", i.nodeBinary, args); err != nil {
+	if err := cli.RunSubCommand(childEnv, i.logger, "identity-init", i.nodeBinary, args); err != nil {
```
Comment (Yawning) [inline go/oasis-test-runner/scenario/e2e/identity_cli.go]: Not your fault but the receiver `i` makes me want to change it to `ident` or something.

Comment (kostko) [reply go/oasis-test-runner/scenario/e2e/identity_cli.go]: Yeah, good idea, done.

File: go/consensus/tendermint/apps/registry/transactions.go (hunk 4 above)
Comment (kostko) [reply go/consensus/tendermint/apps/registry/transactions.go]: I changed the expiry check above to explicitly disallow registering nodes which expire in the current epoch.
